### PR TITLE
Asm to wasm

### PR DESCRIPTION
**Pull request by aseemgarg89**

### CODE REVIEW

File: src/wasm/asm-wasm-builder.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include "third_party/wasm/src/wasm/encoder.h"
+
+namespace v8 {
+namespace internal {
+namespace wasm {
+
+// A Visitor over a CompilationInfo's AST that generates WASM.
+
+class AsmWasmBuilder : public AstVisitor {
+ public:
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.h]: I think you should hide the AstVisitor details in the .cc file.

File: src/wasm/asm-wasm-builder.cc [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+void AsmWasmBuilder::VisitArrayLiteral(ArrayLiteral* expr) {
+  ZoneList<Expression*>* values = expr->values();
+  for (int i = 0; i < values->length(); ++i) {
+    Expression* value = values->at(i);
+    RECURSE(Visit(value));
+  }
+}
+
+void AsmWasmBuilder::VisitAssignment(Assignment* expr) {
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: This won't be a straight recurse, you'll have to pattern match over the LHS of the assignment to know whether you should emit a SetLocal, StoreGlobal, or a StoreMem. For examples of that, have a look at src/compiler/ast-graph-builder.cc.

File: src/wasm/asm-wasm-builder.cc [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+void AsmWasmBuilder::VisitProperty(Property* expr) {
+  RECURSE(Visit(expr->obj()));
+  RECURSE(Visit(expr->key()));
+}
+
+
+void AsmWasmBuilder::VisitCall(Call* expr) {
+  DCHECK(in_function_);
+  current_function_builder_->AddBody(kExprCallFunction, false);
+  RECURSE(Visit(expr->expression()));
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: You'll have to distinguish between direct calls, indirect calls, and calls to imported functions here.

File: src/wasm/asm-wasm-builder.cc [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+  ZoneList<Expression*>* args = expr->arguments();
+  for (int i = 0; i < args->length(); ++i) {
+    Expression* arg = args->at(i);
+    RECURSE(Visit(arg));
+  }
+}
+
+
+void AsmWasmBuilder::VisitCallNew(CallNew* expr) {
+  RECURSE(Visit(expr->expression()));
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Probably shouldn't support these.

File: src/wasm/asm-wasm-builder.cc [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+  ZoneList<Expression*>* args = expr->arguments();
+  for (int i = 0; i < args->length(); ++i) {
+    Expression* arg = args->at(i);
+    RECURSE(Visit(arg));
+  }
+}
+
+
+void AsmWasmBuilder::VisitCallRuntime(CallRuntime* expr) {
+  ZoneList<Expression*>* args = expr->arguments();
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Or this.

File: src/wasm/encoder.cc [hunk 6]
```diff
@@ -90,6 +90,13 @@ void WasmFunctionBuilder::AddBody(
   }
 }
 
+void WasmFunctionBuilder::AddBody(const byte instruction, bool is_local) {
```
Comment (titzer) [inline src/wasm/encoder.cc]: The name AddBody() is weird. You are starting an opcode, yes?
Can you use the WasmOpcode enum here?

File: src/wasm/encoder.h [hunk 7]
```diff
@@ -51,6 +51,7 @@ class WasmFunctionBuilder : public ZoneObject {
   void ReturnType(uint8_t);
```
Comment (titzer) [inline src/wasm/encoder.h]: Please add names to all parameters.

File: test/mjsunit/wasm/asm-wasm.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2015 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+function foo() {
```
Comment (titzer) [inline test/mjsunit/wasm/asm-wasm.js]: "use asm" here?

File: src/wasm/asm-wasm-builder.h (hunk 1 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.h]: done

File: src/wasm/asm-wasm-builder.cc (hunk 2 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: marked the other cases as unreachable for now.

File: src/wasm/asm-wasm-builder.cc (hunk 3 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: marked other use cases as unreachable for now.

File: src/wasm/asm-wasm-builder.cc (hunk 4 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done

File: src/wasm/asm-wasm-builder.cc (hunk 5 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done

File: src/wasm/encoder.cc (hunk 6 above)
Comment (aseemgarg89) [inline src/wasm/encoder.cc]: I want to keep the encoder agnostic to actual wasm op codes. Those are supposed to be encapsulated in the walker. This function is just a special case that can easily take a single byte instead of an array of bytes.

File: src/wasm/encoder.h (hunk 7 above)
Comment (aseemgarg89) [inline src/wasm/encoder.h]: will do as a separate commit

File: test/mjsunit/wasm/asm-wasm.js (hunk 8 above)
Comment (aseemgarg89) [inline test/mjsunit/wasm/asm-wasm.js]: done

File: src/wasm/encoder.cc (hunk 6 above)
Comment (titzer) [inline src/wasm/encoder.cc]: AppendCode(byte opcode, bool add_local_operand) ?

**Comment by titzer** [comment]
LGTM with rename nit

File: src/wasm/encoder.cc (hunk 6 above)
Comment (aseemgarg89) [inline src/wasm/encoder.cc]: done
